scripts/jupyterkernel/kernelsds/src/main/java/com/kernelsds/app/ISystemDsKernel.java

```diff
@@ -0,0 +1,399 @@
+        }
+    }
+
+    /*void analyzeVariables(Hop hop, HashSet<String> inVars, HashSet<String> outVars) {
+        // Check if the hop is a DataOp
+        if (hop instanceof DataOp) {
+            DataOp dataOp = (DataOp) hop;
+            Types.OpOpData opType = dataOp.getOp();
+
+            // If the hop is a TRANSIENTREAD (input variable), add to inVars
+            if (opType == Types.OpOpData.TRANSIENTREAD) {
+                inVars.add(hop.getName());
+            }
+            // If the hop is a TRANSIENTWRITE (output variable), add to outVars
+            else if (opType == Types.OpOpData.TRANSIENTWRITE) {
+                outVars.add(hop.getName());
+            }
+        }
+
+        // Recursively analyze the input hops
+        for (Hop inputHop : hop.getInput()) {
+            analyzeVariables(inputHop, inVars, outVars);
+        }
+    }*/
+    void analyzeStatementBlock(StatementBlock sb, HashSet<String> outVars) {
+        // Extract variable names from 'updated' and 'kill' sets and add them to outVars
+        if (sb.variablesUpdated() != null) {
+            outVars.addAll(sb.variablesUpdated().getVariableNames());
+        }
+    }
+
+    @Override
+    public DisplayData eval(String expr) throws Exception {
+
+        //HashSet<String> inputVariables = new HashSet<>();
+        //HashSet<String> outputVariables = new HashSet<>();
+        String[] outputArray;
+        Set<String> set1 = findInputVariables(expr);
+        Set<String> setoutputs = findInputVariables(expr);
+
+
+
+
+       /*DMLParserWrapper dmlParser = new DMLParserWrapper();
+        DMLProgram dml_program = dmlParser.parse(null,expr,null);
+        DMLTranslator translator = new DMLTranslator(dml_program);
+        translator.validateParseTree(dml_program);
+        translator.liveVariableAnalysis(dml_program);
+        translator.constructHops(dml_program);
+
+
+
+
+        for (StatementBlock sb : dml_program.getStatementBlocks()) {
+            analyzeStatementBlock(sb, outputVariables);
+        }*/
+        int i = 0;
+        for (String str: set1
+        ) {
+            i++;
+            if(this.hashMapForMatrix.get(str) != null){
+                expr =  str + "= read(\"./tmp/doesntexist" + str+ "\", data_type=\"matrix\");\n"+ expr  ;
+            }
+            else{
+                expr =  str + "= read(\"./tmp/doesntexist" + i+ "\", data_type=\"scalar\", value_type=\"string\");\n"+ expr  ;
+            }
+        }
+        //i = 0;
+
+        if (this.hashMap.size() > 0 || this.hashMapForMatrix.size()>0){
+            setoutputs.addAll(extractOutputVariables(expr));
+            setoutputs.addAll(set1);
+            outputArray = setoutputs.toArray(new String[0]);
+
+            for (String out: outputArray
+            ) {
+                //System.out.println("out: "+out);
+                expr =  expr + ";\nwrite("+out+", './tmp/"+out+"');" ;
+            }
```

Comment:
Writing the output objects to disk after each cell is extremely inefficient. We need to utilize the jmlc framework to maintain the variables in-memory.